core/src/main/scala/org/apache/spark/scheduler/TaskSchedulerImpl.scala

```diff
@@ -415,9 +420,55 @@ private[spark] class TaskSchedulerImpl(
             launchedAnyTask |= launchedTaskAtCurrentMaxLocality
           } while (launchedTaskAtCurrentMaxLocality)
         }
+
         if (!launchedAnyTask) {
-          taskSet.abortIfCompletelyBlacklisted(hostToExecutors)
+          taskSet.getCompletelyBlacklistedTaskIfAny(hostToExecutors) match {
+            case Some(taskIndex) => // Returns the taskIndex which was unschedulable
+
+              // If the taskSet is unschedulable we try to find an existing idle blacklisted
+              // executor. If we cannot find one, we abort immediately. Else we kill the idle
```

Comment:
I'm a little worried that the idle condition will be too strict in some scenarios, if there is a large backlog of tasks from another taskset, or whatever the error is, the tasks take a while to fail (eg., you've really got a bad executor, but its not apparent till after network timeouts or something).  Eg. that could happen if you're doing a big join, and while preparing the input on the map-side, one side just has one straggler left but the other side still has a big backlog of tasks.  Or, in a jobserver style situation, and there are always other tasksets coming in.

that said, I don't have any better ideas at the moment, and I still think this is an improvement.

Reply:
By clearing the abort timer as soon as a task is launched we are relaxing this situation. 
If there are large backlog of tasks:
- If we acquire new executors or launch new tasks we will defer the check
- If we cannot acquire new executors and we are running with long running tasks such that no new tasks can be launched and we have less no. of executors compared to max failures, in that case this will end up being harsh. This can happen, but seems more like a very specific edge case.

Reply:
I don't think this is true -- if there is no idle executor here, you abort the taskset immediately, you're not starting any timer, from this case lower down: `case _ => // Abort Immediately`.

I think to do what you described, you would instead need to do something different in that case, like start the same abortTimer, and also set a flag `needToKillIdleExecutor` and then on every call to `resourceOffer`, check that flag and potentially find an executor to kill.  (However I haven't totally thought through that, not sure if it would really work.  again, I'm not saying this has to be addressed now, just thinking this through)